### ISSUE TITLE
memstore saves wrong membership config in `finalize_snapshot_installation`

### ISSUE THREAD

**Issue by lilyball** [railgun-rs/actix-raft#133]
In memstore, in the `MemStore::finalize_snapshot_installation()` method, it throws away the membership config contained in the snapshot, instead saving the membership config contained in its log up through `index` as though that were in the snapshot:

https://github.com/async-raft/async-raft/blob/bcc246ccf8fa2858baf8e07bf760ee6db4d85cf6/memstore/src/lib.rs#L348-L364

As best I can tell, this is a bug and that it should just be using the membership config stored in the snapshot. This code here is very nearly a copy&paste from `MemStore::do_log_compaction()`. It was introduced in 99ac1911f3f9af18a9edc02651f40d521e4c8d60 (part of #59). My best guess here is that the author forgot the snapshot contained a membership config.

Additional evidence for this being a bug is the fact that the code that calls `finalize_snapshot_installation()` then proceeds to request the current membership config from storage and apply it, and the protocol PDF also confirms that the final step to receiving a snapshot is "Reset state machine using snapshot contents (and load snapshot’s cluster configuration)" (Figure 13).

**Comment by drmingdrmer** [comment]
Yes it should check both `ConfigChange` log and `SnapshotPointer` log to extract the correct membership config.

The fix is simple and there is a test that reproduces this bug:
https://github.com/datafuse-extras/async-raft/commit/dba24036cda834e8c970d2561b1ff435afd93165

**Comment by lilyball** [comment]
It should certainly do that during log compaction, that’s a good catch. In the referenced code here it shouldn’t search for a membership config at all though, the `Snapshot` contains one already.

**Comment by lilyball** [comment]
@drmingdrmer I filed #134 about `do_log_compaction` based on your comment.

**Comment by drmingdrmer** [comment]
> It should certainly do that during log compaction, that’s a good catch. In the referenced code here it shouldn’t search for a membership config at all though, the `Snapshot` contains one already.

@lilyball 
Searching for `ConfigChange` is necessary because there could be another member change log that is applied. 
E.g. there could be 4 membership config stored in raft log.
When compacting log, the config change logs that are applied(config-1 and config-2) have to be included in the snapshot.

```
                  .- config-2
                  |      .-config-3
                  |      |     .-config-4
                  v      v     v
+--------------+------+-----+-----+---------> log index
0              |      |     |     `last_log
               |      |     `committed
               |      `applied
               `snapshot-pointer(config-1)
```

**Comment by lilyball** [comment]
@drmingdrmer This ticket is about finalizing a snapshot installation, not about log compaction. The new snapshot that's being finalized already includes whatever membership config is applicable. In this case, a new node will start at

```
+-------------
0
```

And it will receive `snapshot-pointer(config-1)`. But since it throws away the membership there, what it ends up with is

```
+--------------+
0              |
               |
               |
               `snapshot-pointer(new-initial)
```

Note how it's got the wrong membership config, it should have `config-1` but it created a new initial config instead.

Or if we consider a lagging node that looks like the snippet you pasted, and then the node receives a snapshot that replaces eevrything with `config-7` in it, it will end up keeping `config-4` instead and throwing away `config-7`.

**Comment by drmingdrmer** [comment]
@lilyball
Hm... My mistake. Let me clarify it: there are two bugs there:

One of them is what you mentioned: when `finalize_snapshot_installation` it should not read membership from its own log. But instead it should use the membership stored in the `snapshot`(`SnapshotPointer`). But a `SnapshotPointer` log will never be replicated, as far as I recall. It is a **marker** to indicate the leader to switch to `Snapshotting` state.
  

I've fixed this bug in: https://github.com/datafuse-extras/async-raft/commit/447dc11cab51fb3b1925177d13e4dd89f998837b . 

Then this fix exposes the second bug, the one I mentioned above: https://github.com/datafuse-extras/async-raft/commit/dba24036cda834e8c970d2561b1ff435afd93165

**Comment by lilyball** [comment]
Those two commits look good (I didn't read the tests, but the code fixes look correct).

> But a `SnapshotPointer` log will never be replicated, as far as I recall. It is a *marker* to indicate the leader to switch to `Snapshotting` state.

I'm not sure what you mean by this. A `SnapshotPointer` is just the results of either log compaction or receiving a snapshot from the leader. The existence of a `SnapshotPointer` log entry doesn't mean anything for the current snapshot state. AIUI the only time any node _doesn't_ have a `SnapshotPointer` is when it's brand new and hasn't either done a log compaction or received a snapshot. Under normal operation every node in the cluster will end up having a `SnapshotPointer` permanently.

**Comment by drmingdrmer** [comment]
> I'm not sure what you mean by this. A `SnapshotPointer` is just the results of either log compaction or receiving a snapshot from the leader. The existence of a `SnapshotPointer` log entry doesn't mean anything for the current snapshot state. AIUI the only time any node _doesn't_ have a `SnapshotPointer` is when it's brand new and hasn't either done a log compaction or received a snapshot. Under normal operation every node in the cluster will end up having a `SnapshotPointer` permanently.

@lilyball 
Yeah you are totally right. I forgot that the membership is embedded in the snapshot data in this version.

I moved the membership config out of the serialized data then, so that `do_log_compaction` does not need to serialize twice.
I was describing the modified version.
